In [2]:
import torch
print(torch.__version__)

2.10.0+cu128


Creating tensor

In [4]:
#using empty
torch.empty(2,3)
a = torch.empty(2,3)

In [5]:
#check type
type(a)

torch.Tensor

In [6]:
#using zeros
torch.zeros(2,3)

tensor([[0., 0., 0.],
        [0., 0., 0.]])

In [7]:
#using ones
torch.ones(2,3)

tensor([[1., 1., 1.],
        [1., 1., 1.]])

In [10]:
#using random
torch.rand(2,3)

tensor([[0.4443, 0.7519, 0.4692],
        [0.8276, 0.5792, 0.6147]])

In [14]:
#manual seed : it will not show each time different value
torch.manual_seed(100)
torch.rand(2,3)

tensor([[0.1117, 0.8158, 0.2626],
        [0.4839, 0.6765, 0.7539]])

In [15]:
#using tensor
torch.tensor([[1,2,3,4],[5,6,7,8]])

tensor([[1, 2, 3, 4],
        [5, 6, 7, 8]])

## Understanding Tensors

### Subtask:
Explain the fundamental concept of Tensors as generalized multi-dimensional arrays and their significance in data representation.


### 1. Introduction to Tensors

In PyTorch, **Tensors** are the fundamental data structure used to store and manipulate data. They are a generalization of multi-dimensional arrays, similar to NumPy's ndarrays, but with the added capability of being moved to GPUs for accelerated computing.

#### The Hierarchy of Tensors
Tensors are classified by their rank (number of dimensions):

*   **Scalar (0-D Tensor):** A single number (e.g., `5`). Represents a single value.
*   **Vector (1-D Tensor):** An array of numbers (e.g., `[1, 2, 3]`). Represents a list of features or a sequence.
*   **Matrix (2-D Tensor):** A grid of numbers (e.g., a table with rows and columns). Often represents a single grayscale image or a batch of vectors.
*   **Higher-Order Tensors (3-D, 4-D, etc.):** Arrays with more dimensions. For example, a 3-D tensor can represent an RGB image (Height x Width x Channels), and a 4-D tensor can represent a batch of RGB images.

**Reasoning**:
Import the torch library to enable the creation and manipulation of tensor data structures in the subsequent steps.



In [1]:
import torch
print(f'PyTorch version: {torch.__version__}')

PyTorch version: 2.10.0+cpu


## Tensor Properties

### Subtask:
Detail the key attributes of a Tensor: Shape (dimensions), Data Type (dtype), and Device (CPU vs. GPU).


### 2. Key Attributes of a Tensor

Every PyTorch tensor has three essential attributes that define its structure, data representation, and physical location:

1.  **Shape (`.shape` or `.size()`):** Describes the dimensions of the tensor (e.g., a 2x3 matrix has a shape of `[2, 3]`).
2.  **Data Type (`.dtype`):** Specifies the type of data stored in the tensor, such as `torch.float32` (common for weights/gradients) or `torch.int64` (common for labels/indices).
3.  **Device (`.device`):** Indicates whether the tensor is stored in system memory (**CPU**) or on a graphics card (**GPU/CUDA**) for hardware acceleration.

**Reasoning**:
I will create a code cell to initialize a tensor and demonstrate how to access its key attributes: shape, dtype, and device, as well as how to check for GPU availability.



In [16]:
import torch

# Create a sample 2x3 tensor with float32 data type
sample_tensor = torch.tensor([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]], dtype=torch.float32)

# Printing the key attributes
print(f'Tensor Shape: {sample_tensor.shape}')
print(f'Tensor Data Type: {sample_tensor.dtype}')
print(f'Tensor Device: {sample_tensor.device}')

# Check for GPU availability
gpu_available = torch.cuda.is_available()
print(f'\nIs GPU (CUDA) available? {gpu_available}')

if gpu_available:
    # Example of moving a tensor to the GPU
    gpu_tensor = sample_tensor.to('cuda')
    print(f'Moved Tensor Device: {gpu_tensor.device}')
else:
    print('Note: Tensors can be moved to GPU using .to(\'cuda\') if a GPU is available.')

Tensor Shape: torch.Size([2, 3])
Tensor Data Type: torch.float32
Tensor Device: cpu

Is GPU (CUDA) available? True
Moved Tensor Device: cuda:0


## PyTorch Tensor Operations

### Subtask:
Demonstrate practical PyTorch code for creating tensors, indexing/slicing, reshaping, and performing mathematical operations.


### 3. Practical Tensor Operations

To effectively use PyTorch for deep learning, one must master several core operations:

*   **Creation:** Initializing tensors with specific values (zeros, ones, or random numbers) to set up weights or dummy data.
*   **Indexing and Slicing:** Extracting specific portions of data, similar to Python lists or NumPy arrays.
*   **Reshaping:** Changing the dimensions of a tensor (e.g., flattening an image) while keeping the total number of elements constant.
*   **Mathematical Operations:** Performing the computations that drive neural network training, from element-wise arithmetic to matrix multiplications.

**Reasoning**:
I will now provide a code cell that demonstrates how to create tensors using common functions like zeros, ones, and random values, alongside examples of indexing, slicing, and reshaping.



In [17]:
import torch

# 1. Tensor Creation
zeros = torch.zeros((2, 3))
ones = torch.ones((2, 3))
random_tensor = torch.rand((3, 4))

print(f'Zeros:\n{zeros}')
print(f'Ones:\n{ones}')
print(f'Random (0-1):\n{random_tensor}\n')

# 2. Indexing and Slicing
data = torch.tensor([[10, 20, 30], [40, 50, 60], [70, 80, 90]])
print(f'Original Matrix:\n{data}')
print(f'Element at row 0, col 1: {data[0, 1]}')
print(f'First two rows:\n{data[:2, :]}')
print(f'Last column: {data[:, -1]}\n')

# 3. Reshaping
# Reshaping a 3x3 (9 elements) to 1x9 or 9x1
reshaped = data.view(1, 9)
print(f'Reshaped (1x9):\n{reshaped}')
print(f'Original elements: {data.numel()}, Reshaped elements: {reshaped.numel()}')

Zeros:
tensor([[0., 0., 0.],
        [0., 0., 0.]])
Ones:
tensor([[1., 1., 1.],
        [1., 1., 1.]])
Random (0-1):
tensor([[0.2627, 0.0428, 0.2080, 0.1180],
        [0.1217, 0.7356, 0.7118, 0.7876],
        [0.4183, 0.9014, 0.9969, 0.7565]])

Original Matrix:
tensor([[10, 20, 30],
        [40, 50, 60],
        [70, 80, 90]])
Element at row 0, col 1: 20
First two rows:
tensor([[10, 20, 30],
        [40, 50, 60]])
Last column: tensor([30, 60, 90])

Reshaped (1x9):
tensor([[10, 20, 30, 40, 50, 60, 70, 80, 90]])
Original elements: 9, Reshaped elements: 9


**Reasoning**:
I will add a new code cell to demonstrate basic element-wise mathematical operations and matrix multiplication using torch.matmul and the @ operator as required by the subtask instructions.



In [18]:
import torch

# Define two sample tensors for operations
a = torch.tensor([[1, 2], [3, 4]])
b = torch.tensor([[5, 6], [7, 8]])

# 1. Element-wise Addition
sum_result = a + b
print(f'Element-wise Addition:\n{sum_result}')

# 2. Element-wise Multiplication
elementwise_prod = a * b
print(f'\nElement-wise Multiplication:\n{elementwise_prod}')

# 3. Matrix Multiplication (Dot Product)
# Method 1: Using torch.matmul()
matmul_result = torch.matmul(a, b)
# Method 2: Using the @ operator
matmul_operator = a @ b

print(f'\nMatrix Multiplication (using @ or matmul):\n{matmul_result}')
print(f'Check if results are equal: {torch.equal(matmul_result, matmul_operator)}')

Element-wise Addition:
tensor([[ 6,  8],
        [10, 12]])

Element-wise Multiplication:
tensor([[ 5, 12],
        [21, 32]])

Matrix Multiplication (using @ or matmul):
tensor([[19, 22],
        [43, 50]])
Check if results are equal: True


## Tensors in Deep Learning

### Subtask:
Explain how Tensors are used to represent inputs, model weights/parameters, and how they store gradients for training.


### 4. Tensors in Deep Learning

In the context of deep learning, Tensors serve three primary functions:

*   **Data Representation:** Input data is fed into models as tensors. For example, a batch of 32 RGB images of size 224x224 would be a 4D tensor with shape `[32, 3, 224, 224]`.
*   **Model Parameters:** Neural network weights and biases are stored as tensors. These tensors are updated during the training process to minimize the loss function.
*   **Automatic Differentiation:** PyTorch tensors can track their computational history. When a tensor's `requires_grad` attribute is set to `True`, PyTorch records operations performed on it, allowing for the automatic calculation of gradients (derivatives) via the `.backward()` method. These gradients are stored in the `.grad` attribute of the tensor and are used by optimizers to update weights.

**Reasoning**:
I will create a code cell to demonstrate how tensors are used as model weights with the `requires_grad=True` attribute, perform a simple forward pass, and compute gradients using `.backward()`.



In [19]:
import torch

# 1. Representing Model Parameters (Weights)
# We set requires_grad=True to tell PyTorch to track operations for gradient calculation
weight = torch.tensor([2.0], requires_grad=True)
bias = torch.tensor([1.0], requires_grad=True)

# 2. Representing Input Data
x = torch.tensor([1.5])

# 3. Simple Operation (Forward Pass: y = weight * x + bias)
y = weight * x + bias

# 4. Automatic Differentiation (Backward Pass)
# This computes the derivative of y with respect to weight and bias
y.backward()

print(f'Input (x): {x.item()}')
print(f'Weight: {weight.item()}')
print(f'Bias: {bias.item()}')
print(f'Output (y): {y.item()}')

# 5. Accessing Gradients
print(f'\nGradient of Weight (dy/dweight): {weight.grad.item()}')
print(f'Gradient of Bias (dy/dbias): {bias.grad.item()}')

print('\nNote: The gradient of weight is 1.5 because dy/dweight = x.')

Input (x): 1.5
Weight: 2.0
Bias: 1.0
Output (y): 4.0

Gradient of Weight (dy/dweight): 1.5
Gradient of Bias (dy/dbias): 1.0

Note: The gradient of weight is 1.5 because dy/dweight = x.


## Benefits and Helpfulness

### Subtask:
Explain the advantages of using PyTorch Tensors, specifically focusing on GPU acceleration and Automatic Differentiation (Autograd).


### The Advantages of PyTorch Tensors

PyTorch Tensors provide two critical advantages over standard arrays (like Python lists or NumPy arrays) that make them indispensable for deep learning:

1. **GPU Acceleration (Parallelism):** While CPUs are optimized for complex sequential tasks, GPUs consist of thousands of smaller cores designed for massive parallel processing. Since deep learning involves large-scale matrix and vector operations, Tensors can be offloaded to the GPU to achieve significant speedups.
2. **Automatic Differentiation (Autograd):** PyTorch's `autograd` engine automatically builds a dynamic computational graph as you perform operations. This graph tracks every transformation, allowing PyTorch to compute complex gradients using the chain rule with a single call to `.backward()`, which is essential for optimizing neural network weights.

**Reasoning**:
I will create a code cell that demonstrates GPU availability and a clear example of Autograd's computational graph using a trigonometric function to show automatic chain-rule differentiation.



In [1]:
import torch
import time

# 1. GPU Acceleration Check
device = "cuda" if torch.cuda.is_available() else "cpu"
print(f"Current computation device: {device.upper()}")

# Simulating a large operation to discuss acceleration
size = 5000
x = torch.randn(size, size)

start_time = time.time()
y = x @ x
end_time = time.time()
print(f"Time taken for {size}x{size} matrix multiplication on {device}: {end_time - start_time:.4f} seconds")

if device == "cpu":
    print("Note: Matrix operations are significantly faster on GPU due to massive parallelism.\n")

# 2. Automatic Differentiation (Autograd) in Action
# Let's use a function: y = sin(x^2)
# The derivative should be: dy/dx = cos(x^2) * 2x

x_val = torch.tensor([1.5], requires_grad=True)
y = torch.sin(x_val**2)

# Trigger the backward pass to compute gradients
y.backward()

# Theoretical gradient calculation for verification
with torch.no_grad():
    expected_grad = torch.cos(x_val**2) * 2 * x_val

print("--- Autograd Computational Graph Example ---")
print(f"Input x: {x_val.item()}")
print(f"Function y = sin(x^2): {y.item():.4f}")
print(f"PyTorch Calculated Gradient (dy/dx): {x_val.grad.item():.4f}")
print(f"Manual Chain Rule Calculation: {expected_grad.item():.4f}")
print("\nAutograd successfully handled the trigonometric and power function chain rule!")

Current computation device: CUDA
Time taken for 5000x5000 matrix multiplication on cuda: 1.9020 seconds
--- Autograd Computational Graph Example ---
Input x: 1.5
Function y = sin(x^2): 0.7781
PyTorch Calculated Gradient (dy/dx): -1.8845
Manual Chain Rule Calculation: -1.8845

Autograd successfully handled the trigonometric and power function chain rule!


## Final Task

### Subtask:
Provide a comprehensive summary of the topics discussed and answer any final questions about their utility in PyTorch.


## Summary:

This summary covers the fundamental concepts, properties, and operations of PyTorch Tensors, as well as their specialized role in deep learning.

### Q&A

**What are the primary advantages of PyTorch Tensors over standard arrays like NumPy?**
The two critical advantages are **GPU Acceleration** and **Automatic Differentiation (Autograd)**. Tensors can be moved to GPUs (using `.to('cuda')`) to perform massive parallel computations, which is essential for large-scale matrix operations in neural networks. Additionally, the `autograd` engine tracks computational history, allowing for the automatic calculation of complex gradients via the `.backward()` method, eliminating the need for manual calculus during model training.

**How do Tensors represent different types of data in a deep learning pipeline?**
Tensors use a hierarchy of ranks:
*   **0-D (Scalars):** Single values like loss or accuracy.
*   **1-D (Vectors):** Lists of features or sequences.
*   **2-D (Matrices):** Tabular data or single-channel (grayscale) images.
*   **Higher-Order (3-D, 4-D+):** RGB images (Height x Width x Channels) or batches of images (Batch x Channels x Height x Width).

---

### Data Analysis Key Findings

*   **Tensor Attributes:** Every tensor is defined by its **Shape** (structural dimensions), **Data Type** (e.g., `torch.float32` for precision), and **Device** (CPU or CUDA/GPU).
*   **Structural Flexibility:** The `.view()` method allows for seamless reshaping of data (e.g., transforming a 3x3 matrix into a 1x9 vector) as long as the total number of elements (`.numel()`) remains constant.
*   **Mathematical Operations:** PyTorch supports both element-wise operations (addition, multiplication) and matrix-level operations. Matrix multiplication can be performed using `torch.matmul()` or the `@` operator.
*   **Gradient Tracking:** By setting `requires_grad=True`, PyTorch successfully tracks operations. In a test case for the function $y = \sin(x^2)$, the `autograd` engine precisely calculated the gradient ($\approx -1.8845$), matching manual chain-rule derivations.
*   **Hardware Utility:** Code execution demonstrated that even when a GPU is unavailable, the `.to('device')` logic provides a portable way to write code that scales from local CPUs to high-performance cloud GPUs.

---

### Insights or Next Steps

*   **Standardize Data Types:** When building models, ensure input tensors and model weights share the same `dtype` (typically `float32`) and are on the same `device` to avoid runtime errors.
*   **Leverage Vectorization:** Use matrix operations (like `@`) instead of manual loops for computations; this allows PyTorch to utilize underlying C++/CUDA optimizations for significantly faster execution.
